Set, Union, & Intersection iterators.

willhite opened the pull request.

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,213 @@
+	// AdvanceTo(v) returns:
+	// * nil if there are no objects >= v.
+	// * the next value in the iteration if  v <= curValue
+	// * the first value > v if v is not in the set, or nil if there is no value > v
```
The second part of this point "nil if there is no value > v" contradicts the first point "nil if there are no objects >= v".

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,213 @@
+	// Next returns nil, when no objects remain
+	Next() Value
+
+	// AdvanceTo(v) returns:
```
AdvanceTo ... also advances the cursor right? So this comment could be "AdvanceTo advances the cursor to the next value >= `v`".

kalman replied:
Followed by "The return value is ...?"
And for that matter what state is the cursor in after this point?

kalman replied:
Another independent comment here: looking at the tests, the behavior here does seem reasonable, though the method might be better named similar to `NextFrom`. `AdvanceTo` sounds like it does just take the iterator to a position such that the next call to `Next` returns that value.

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,213 @@
+	cursor *sequenceCursor
+}
+
+func (s Set) Iterator() SetIterator {
```
I think it's better to define this in set.go

kalman replied:
^

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,213 @@
+	return &setIterator{s: s, cursor: nil}
+}
+
+func (si *setIterator) Next() (v Value) {
```
the `v` in `(v Value)` isn't used, likewise the `res` in `(res Value)` below... and almost all of them?

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,213 @@
+	return nil
+}
+
+func (si *setIterator) AdvanceTo(v Value) (res Value) {
```
It seems like a simpler to understand API here would be for AdvanceTo to advance the cursor to `v`, and return the Value it's now on. The caller can then call `Next` if they want. It would also make the code simpler I think.

willhite replied:
The problem is that these iterators can be nested, which means shifting the cursor around can mean shifting lots of cursors. I was looking for a semantic that would be universally applicable across all 3 types of iterators and constrain the user to only do sensible things.

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,213 @@
+	return nil
+}
+
+// iterState contains iterator and it's current value
```
You could also just put a Value() method on Iterator? a little bit weird for an iterator to have such a method, but I could see it being useful in other circumstances.

willhite replied:
I did think about this. It would allow AdvanceTo(v) to return the current value when v == current value. But I opted to keep it as streamlined as possible. I'm open to changing that.

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,213 @@
+}
+
+func NewUnionIterator(iterA, iterB SetIterator) SetIterator {
+	if iterA == nil || iterB == nil {
```
Why not just `d.Chk.NotNil(iterA); d.Chk.NotNil(iterB)`?

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,213 @@
+		}
+		return iterA
+	}
+	a := iterState{i: iterA, v: iterA.Next()}
```
I also think having Value on SetIterator would make this simpler, you wouldn't need to call Next I think

willhite replied:
Whatever gets supported on SetIterator would need to be supported by Union and Intersection as well.

kalman replied:
Fair enough, you've convinced me.

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,213 @@
+}
+
+func (u *UnionIterator) Next() (res Value) {
+	i := compareValue(u.aState.v, u.bState.v)
```
`switch (compareValue)`?

kalman replied:
also switch here, etc

kalman commented on go/types/set_iterator_test.go:
```diff
@@ -0,0 +1,174 @@
+	vs := ValueSlice{}
+	for v := i.Next(); v != nil; v = i.Next() {
+		vs = append(vs, v)
+	}
```
Pattern is repeated a lot, how about a method like `iterToSlice(SetIterator) ValueSlice`? then just `vs := iterToSlice(i)` here.

kalman commented on go/types/set_iterator_test.go:
```diff
@@ -0,0 +1,174 @@
+
+	i = s.Iterator()
+	assert.Panics(func() { i.AdvanceTo(nil) })
+	assert.Equal(Number(0), i.AdvanceTo(Number(-20)))
```
It would be good to test what happens when calling Next after AdvanceTo. E.g. presumably `i.AdvanceTo(5); iterToSlice(i) => [6, 10, 15]`?

kalman commented:
Back to you

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,213 @@
+	if u.bState.v == nil {
+		return u.aState.Next()
+	}
+	if u.aState.v.Less(u.bState.v) {
```
use `compareValue`?

kalman replied:
^

I.e. would this work
```
...
if !didAdvance {
  return u.Next()
}
switch compareValue(u.aState.v, u.bState.v) {
  case -1:
    return u.bState.Next()
  case 1:
    return u.aState.Next()
  case 0:
    u.aState.Next()
    return u.bState.Next()
}
panic("unreachable")
```

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,213 @@
+
+func (u *IntersectionIterator) Next() (res Value) {
+	cnt := 0
+	for i := compareValue(u.aState.v, u.bState.v); i != 0; i = compareValue(u.aState.v, u.bState.v) {
```
Perhaps nicer:
```
for {
  switch compareValue(...) {
    case: -1: ...
    case: 1: ...
    case: 0
      break
  }
}
```

willhite commented:
@kalman PTAL, I didn't change the name of AdvanceTo(). I'm not convinced there's anything better. Other than that, I think I've addressed all your comments.

arv commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,213 @@
+
+import "github.com/attic-labs/noms/go/d"
+
+type SetIterator interface {
```
Please document all new public API.

arv replied:
Should this just be `Iterator`? I don't see what is specific to Set here?

kalman replied:
Abstractly yes, but practically no:

A Next() function makes sense to return a Value, but only for Sets and Lists; Maps would return (Value, Value).

And AdvanceTo isn't possible on Lists.

So this Iterator could only be used (in our current data structures) on a Set.

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,208 @@
+	Next() Value
+
+	// AdvanceTo(v) advances the cursor to a new position based on the following rules and returns the value at that position.
+	// * if v <= curValue, advances to the next value in the iteration.
```
More thoughts:

The behaviour that if the current value < v then it advances is strange. If I have [1,2,3,4,5], current value is 3, and I "advance to" 1, then the iterator is now at 4. That's counterintuitive to me.

I still think this would ideally be either better named, or have a different interface, or different behavior depending on which is most sensible.

E.g. what if the comment were actually "like Next(), but returns a value > `v`", which I think is what it's doing at the moment? That makes a lot more sense to me (and the name NextFrom evokes that more than AdvanceTo, but don't feel particularly strongly about that).

kalman replied:
I guess it should be "returns a value >= v".

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,208 @@
+		return curValue
+	}
+
+	if v.Less(curValue) {
```
Should this be <= not <? I.e. !curValue.Less(v)

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,208 @@
+}
+
+func (st *iterState) AdvanceTo(v Value) Value {
+	if st.v == nil || v == nil {
```
would prefer d.Chk.NotNil(v) and then just `if st.v == nil { return nil }`

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,208 @@
+		return u.bState.Next()
+	}
+
+	return nil
```
panic("unreachable")?

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,208 @@
+
+func (u *IntersectionIterator) AdvanceTo(v Value) Value {
+	d.Chk.NotNil(v)
+	if u.aState.v != nil && !v.Less(u.aState.v) {
```
I've seen this pattern a few times now, maybe everything should just be using `compareValue` which both documents the behavior of nil, and does the check properly.

Up to you.

kalman replied:
i.e. `if compareValue(v, u.aState.v) == -1 {`

kalman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,208 @@
+}
+
+// considers nil max value, return -1 if v1 < v2, 0 if v1 == v2, 1 if v1 > v2
+func compareValue(v1, v2 Value) int {
```
IMO this function would be more explicit like

```
if v1 == nil && v2 == nil {
  return 0
}
if v2 == nil || v1.Less(v2) {
  return -1
}
if v1 == nil || v2.Less(v1) {
  return 1
}
return 0
```

up to you.

kalman commented:
lgtm with a bunch more comments

kalman commented on go/types/set_iterator.go:
```diff
@@ -113,45 +116,45 @@ func (u *UnionIterator) Next() Value {
 		return u.bState.Next()
 	}
-
+	d.Chk.Fail("Unreachable code")
```
I would just panic (don't see the point in using our test-assert wrappers here)

kalman commented on go/types/set_iterator_test.go:
```diff
@@ -161,3 +163,146 @@ func TestCombinationIterator(t *testing.T) {
 	assert.True(vs.Equals(expectedRes), "Expected: %v != actual: %v", expectedRes, vs)
 }
+
+var callCount = 0
```
can you put this on UnionTestIterator/IntersectionTestIterator instead?

kalman replied:
or a pointer

kalman commented:
lgtm

aboodman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,207 @@
+// SetIterator defines methods that can be used to efficiently iterate through a set in 'Noms-defined'
+// sorted order.
+type SetIterator interface {
+	// Next returns subsequent values from a set. It returns nil, when no objects remain
```
Missing period at end of comment.

aboodman commented on go/types/set_iterator.go:
```diff
@@ -0,0 +1,207 @@
+	Next() Value
+
+	// NextFrom(v) advances the cursor to a new position based on the following rules and returns the value at that position.
+	// * if v <= curValue, advances to the next value in the iteration.
```
I don't understand this case at all -- first of all 'curValue' is not defined in this interface anywhere, so I don't know what it refers to...

willhite replied:
Would this work?
+	// * if v <= last returned value of the iterator, the next value in the iteration is returned.
